### PR TITLE
Add a rank-k update gufunc (i.e. SYRK)

### PR DESCRIPTION
**Pull request by grlee77**
### Overview

This PR adds a new function `update_rankk` which implements a `gufunc` for the BLAS3 functions `ssyrk` and `dsyrk`.

Complex valued support (`csyrk` and `zsyrk`) can be added in a follow-up PR later if needed.

The `transpose_type` and `UPLO` argument names match the names used by other `gulinalg` functions.

### Additional Info

[Reference documentation for syrk](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/blas-and-sparse-blas-routines/blas-routines/blas-level-3-routines/cblas-syrk.html)

Note: currently the default values of 1.0 for `alpha` and `beta` parameters are used as in other `gulinalg` functions.

### CODE REVIEW

**Comment by grlee77** [comment]
This first commit here is from #1, the rest are specific to adding syrk

**Comment by grlee77** [comment]
@teoliphant, @Carreau: This PR corresponds to https://github.com/Quansight/deshaw/issues/102. It should be working now if either of you can do a quick review

I tested locally with the following two environments on linux:
- Python 2.7.15 + NumPy 1.16.5 + libopenblas 0.3.12
- Python 3.8.5 + NumPy 1.19.1 + libopenblas 0.3.10

This repository is still configured to use `nose`, so if it is installed you can run the tests using:
```
python gulinalg/tests/test_gufunc_general.py
```
or via pytest
```
pytest gulinalg/tests/test_gufunc_general.py -v
```

Also, who would have access to enable Travis-CI for this repository? I don't see Quansight listed under my organizations on travis-ci.org, so I don't think I currently have the permissions needed to do it.

File: gulinalg/gufunc_general.py [hunk 1]
```diff
@@ -359,3 +359,78 @@ def update_rank1(a, b, c, conjugate=True, **kwargs):
...
+    Numpy broadcasting rules apply.
+
+    Implemented for single, double. Numpy conversion
+    rules apply.
+
+    Rank-k update is computed using BLAS _syrk functions.
+
+    Examples
+    --------
+    TODO
```
Comment (Carreau) [inline gulinalg/gufunc_general.py:404]: Want to fill that or remove this example ?

File: gulinalg/gufunc_general.py [hunk 2]
```diff
@@ -359,3 +359,78 @@ def update_rank1(a, b, c, conjugate=True, **kwargs):
         gufunc = _impl.update_rank1
 
     return gufunc(a, b, c, **kwargs)
+
+
+def update_rankk(a, c, UPLO='U', transpose_type='N', **kwargs):
```
Comment (Carreau) [inline gulinalg/gufunc_general.py:364]: Are kwargs necessary ? Here and in the return at the end ?

File: gulinalg/src/gulinalg.c.src [hunk 3]
```diff
@@ -2095,6 +2107,158 @@ static void
...
+
+    if (trans == 'N') {
+        n = (fortran_int) dimensions[0];
+        k = (fortran_int) dimensions[1];
+    }
+    else {
+        /* TODO: fix trans != N case. currently this branch is never called
+           because gufunc_general.py does the transpose so that this function
+           is always called with trans == 'N'
+         */
```
Comment (Carreau) [inline gulinalg/src/gulinalg.c.src:2198]: 👍

**Review by Carreau — APPROVED** [review]
Looks good to me, though I'm no usually writing C, so you may want someone else pairs of eyes.

**Comment by Carreau** [comment]
I've also requested github access for travis-ci, and admin needs to approve.

File: gulinalg/gufunc_general.py (hunk 2 above)
Comment (grlee77) [reply gulinalg/gufunc_general.py:364]: Yes, these functions support all of the general ufunc options as documented here: https://numpy.org/doc/stable/reference/ufuncs.html#ufunc

I don't currently have specific tests exercising these, but I did test `out` and `dtype` kwargs locally just to make sure those were working as expected.

File: gulinalg/gufunc_general.py (hunk 1 above)
Comment (grlee77) [reply gulinalg/gufunc_general.py:404]: whoops. I will add an simple example

**Comment by grlee77** [comment]
Travis now looks good aside from a handful of test failures unrelated to this PR. I can probably just change `atol` from 0 to something like 1e-14 for these `TestSolveTriangular` tests.

**Comment by grlee77** [comment]
I fixed up the doctests and updated the Travis config so that everything is green.

**Comment by grlee77** [comment]
Thanks for reviewing @Carreau. I plan to merge tomorrow morning if there are no further comments before then.

**Pull request merged**